src/services/implementation/ClientServiceImpl.java

```diff
@@ -1,73 +1,11 @@
 package services.implementation;
 
-import enums.StatusEnum;
-import exceptions.services.EntityNotFoundException;
-import exceptions.services.IdNullPointerException;
-import exceptions.services.NullEntityException;
-import java.time.Instant;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Optional;
-import java.util.stream.Collectors;
+import entities.Client;
 import services.interfaces.ClientService;
-import util.ClientServiceUtility;
-import util.UserUtil;
-
-public class ClientServiceImpl implements ClientService {
-  private final List<entities.Client> clients;
-  private Long currentId;
+import util.ClientUtil;
 
+public class ClientServiceImpl extends BaseServiceImpl<Client, Long> implements ClientService {
   public ClientServiceImpl() {
-    this.clients = new ArrayList<>();
-    this.currentId = 0L;
-  }
-
-  @Override
-  public Optional<entities.Client> insert(entities.Client client) {
-    client.setId(this.currentId);
-    client.setCode(ClientServiceUtility.createCode(this.currentId));
-    client.setStatus(StatusEnum.ACTIVE);
-    client.setCreatedAt(Instant.now());
-    client.setUpdatedAt(Instant.now());
-    this.currentId = this.currentId + 1;
-    this.clients.add(client);
-    return Optional.of(client);
-  }
-
-  @Override
-  public Optional<entities.Client> findById(Long id) {
-    Optional.ofNullable(id).orElseThrow(IdNullPointerException::new);
-    var client =
-        this.clients.stream()
-            .filter(c -> id.equals(c.getId()) && !(c.getStatus() == StatusEnum.DELETED))
-            .findAny()
-            .orElseThrow(() -> new EntityNotFoundException(entities.Client.class, id));
-    return Optional.ofNullable(client);
-  }
-
-  @Override
-  public Optional<entities.Client> delete(Long id) {
-    Optional.ofNullable(id).orElseThrow(IdNullPointerException::new);
-    var client = this.findById(id).get();
-    client.setStatus(StatusEnum.DELETED);
-    client.setUpdatedAt(Instant.now());
-    return Optional.of(client);
-  }
-
-  @Override
-  public Optional<entities.Client> update(entities.Client entity) {
-    Optional.ofNullable(entity).orElseThrow(NullEntityException::new);
-    Optional.ofNullable(entity.getId()).orElseThrow(IdNullPointerException::new);
-    var client = this.findById(entity.getId()).get();
-    entity.setUpdatedAt(Instant.now());
-    UserUtil.fetchUser(client, entity);
-    return Optional.of(client);
-  }
-
-  @Override
-  public List<entities.Client> findAll() {
-    return this.clients.stream()
-        .filter(c -> c.getStatus() != StatusEnum.DELETED)
-        .collect(Collectors.toList());
+    super(0L, Client.class, currentId -> currentId + 1, ClientUtil::fetchClient);
```

Comment:
Within base service util create 2 maps:

1. To map default key with its generator function. Ex: `public static final Map<Class<?>, BiFunction<?, ? , ?>> ENTITY_TO_FETCH_FUNCTION_MAP = Map.of(Client.class, ClientUtil::fetchClient, ....);` 
2. To map the entity class with its fetch function. Ex: `public static final Map<Class<?>, BiFunction<?, ? , ?>> ENTITY_KEY_TO_GENERATOR_FUNCTION_MAP = Map.of(Long.class, currentId -> currentId + 1);`

And then create a functions to get the values of each map respectively given a key.

Reply:
What i did is create the map, and BaseServiceImpl simply calls it with .get(Entity.class) and it returns the BiFunction that fetch the correct entity